### ISSUE TITLE
Faulty interface design of single web crawler running in multiple threads

### ISSUE THREAD

**Issue by geskill** [yasserg/crawler4j#108]
I can specify with the number of crawlers the amount of concurrent threads to be run via the `CrawlController`. Every crawler thread uses the same `CrawlConfig`.

This is good for some settings but bad for some other settings. For example the `resumableCrawling`, `maxDepthOfCrawling` or `maxPagesToFetch` is globally useful for all threads, but proxy settings (`proxyHost`, `proxyUsername`, …) or `userAgentString` should be definable per each thread instance and not the same value for all concurrent threads, as i understand your settings interface so far.

In order to realize the desired setup, i have to define for the preferred amount of threads a separate `CrawlController` that can _then_ use a separate `CrawlConfig`. However i am not sure, if i can use the same storage folder, resumable crawling and whats most important if the constrollers share their work-list, which i assume are not do.

This finally brings be to the point that the interface for multiple crawlers is designed wrong, not to say broken.

Some pull requests that try to workaround this design error: https://github.com/yasserg/crawler4j/pull/80 https://github.com/yasserg/crawler4j/pull/57

Probably the best solution to fix this, it to distinguish between globally settings `CrawlConfig` and _per thread_ settings that inherit from the global settings but can be overwritten. Furthermore, the  _per thread_ settings have less options, limiting to proxy settings, user agent, delay ... In class hierarchy `CrawlConfig` would inherit from the basic _per thread_ settings.

**Comment by Chaiavi** [comment]
I see your point.

But, as this seems as a big refactor, what do you think about the following:
Maintain the same code logic as it is now but have an option to configure
multiple user-agents & proxy settings, so each thread can take a random
user agent and proxy and do its thing.


What do you say, will we thus gain the same results ?

**Comment by geskill** [comment]
The big part is to changing the `PageFetcher` to be initialized with variable parameters depending on the thread instance. Currently the `HttpClient` is one instance that is created (inside the PageFetcher) and will be used from every thread. (This is partly already done here: [PageFetcher.java](https://github.com/Bouki/crawler4j/blob/master/src/main/java/edu/uci/ics/crawler4j/fetcher/PageFetcher.java))

Once changing is done, i think it will be easier to create a class `CrawlConfigBase` that is a member of the `WebCrawler`, instead of a function from the `PageFetcher`, which asks for some setup for thread x (as far as i understand you).

However your setup would not break the `CrawlController.start` functions and you are completely correct either way involves multiple code changes.

The crawler is just not designed to work in some lets say stealth mode, when it comes to crawling on a single webpage.
